client/src/controllers/SearchController.ts

```diff
@@ -0,0 +1,185 @@
+import { debounce } from '../utils/debounce';
+
+declare global {
+  interface Window {
+    headerSearch?: { targetOutput: string; url: string };
+  }
+}
+
+/**
+ * Allow for an element to trigger an async search query that will
+ * path the results into a results DOM container. The search query
+ * input can be the controlled element or the containing form.
+ */
+export class SearchController extends Controller<
+  HTMLFormElement | HTMLInputElement
+> {
+  static targets = ['input'];
+
+  static values = {
+    loading: { default: false, type: Boolean },
+    pageParam: { default: 'p', type: String },
+    queryParam: { default: 'q', type: String },
+    results: { default: '#results', type: String },
+    url: { default: '', type: String },
+    wait: { default: 200, type: Number },
+  };
+
+  abortController?: AbortController;
+  hasInputTarget: boolean;
+  hasResultsValue: boolean;
+  hasUrlValue: boolean;
+  /** The related icon element to attach the spinner to */
+  icon: SVGUseElement | null;
+  /** The default icon name that may be present on load */
+  iconName: string | null;
+  inputTarget: HTMLInputElement;
+  loadingValue: boolean;
+  pageParamValue: any;
+  queryParamValue: any;
+  /** Element that receives the results HTML output */
+  resultsElement: HTMLElement;
+  resultsValue: string;
+  /** Debounced function to fetch results & patch into the DOM */
+  search: { (...args: any[]): void; cancel(): void };
+  urlValue?: string;
+  waitValue: number;
+
+  connect() {
+    const formContainer = this.hasInputTarget
+      ? this.inputTarget.form
+      : this.element;
+
+    // support legacy window global approach (add warning in the future)
+    const { url, targetOutput } = window.headerSearch || {};
+    this.resultsValue = targetOutput || this.resultsValue;
+    this.urlValue =
+      url || this.urlValue || formContainer?.getAttribute('action') || '';
+    this.icon = null;
+
+    const resultsElement = document.querySelector(this.resultsValue);
+
+    const hasValidResultsElement =
+      resultsElement && resultsElement instanceof HTMLElement;
+    const hasValidUrlValue = !!this.urlValue;
+
+    if (!hasValidResultsElement || !hasValidUrlValue) {
+      throw new Error(
+        [
+          !hasValidUrlValue && 'Cannot find search URL value',
+          !hasValidResultsElement &&
+            `Cannot find results element at "${this.resultsValue}"`,
+        ]
+          .filter(Boolean)
+          .join(', '),
+      );
+    }
+
+    this.resultsElement = resultsElement as HTMLElement;
+
+    // set up icons
+    const iconContainer = (
+      this.hasInputTarget ? this.inputTarget : this.element
+    ).parentElement;
+
+    this.icon = iconContainer?.querySelector('use') || null;
+    this.iconName = this.icon?.getAttribute('href') || null;
+    this.loadingValue = false;
+
+    // set up debounced search action method
+    this.search = debounce(this.runSearch.bind(this), this.waitValue);
```

Comment:
```suggestion
    this.update = debounce(this.search.bind(this), this.waitValue);
```
Update may be a better method name